### DISCUSSION TITLE
Memory Map and Pico6502 Block Diagrams

### DISCUSSION THREAD

**Discussion by netzerohero** [picocomputer/community#64] [Q&A]
For what it's worth, I've been putting together a memory-map (for the 6502-side) and captured a few HW-architecture Block-diagrams from @rumbledethumps videos.  I'm doing this to improve my understanding of both the CC65-environment, the EhBASIC environment as well as the supporting hardware.

Unsure the best way to share (or version-control) this type of documentation, so just attaching the spreadsheet here.

Comments welcome.

[Pico6502_BlockDiagrams_RAM_(20231216_1744).xlsx](https://github.com/picocomputer/community/files/13694155/Pico6502_BlockDiagrams_RAM_.20231216_1744.xlsx)
[updated: 20231216_1744]

[Pico6502_BlockDiagrams_RAM_(20240512_1258).xlsx](https://github.com/user-attachments/files/17612094/Pico6502_BlockDiagrams_RAM_.20240512_1258.xlsx)
[updated with info from: 20240512_1258]

**Comment by linuxplayground** [comment]
I typically think of the 256 byte blocks of ram as pages which I name according to the high byte.  So page 0 is zeropage and page FF is where the vectors live etc.

ehbasic has extensive use of Zeropage.  In my personal project have it laid out in two segments.  One for stuff that must live at the bottom of zero page and another that can be relocatable within zero page.  See: https://github.com/linuxplayground/6502-retro-banked-v3/blob/main/rom/basic/basic.s

The usage then ends up like this:  (NOTE: THIS IS NOT PICOCOMPUTER STUFF, but used here to illustrate my points)

```
Name                   Start     End    Size  Align
----------------------------------------------------
BASICZP0              000000  000012  000013  00001
KERNZP                000013  000018  000006  00001
BASICZP               00001B  0000A6  00008C  00001
CODE                  00C000  00E949  00294A  00001
VECTORS               00FFFA  00FFFF  000006  00001
```
As you can see, the last address of ZEROPAGE used by this particular ROM is at A6.  So addresses A7 - FF are free for the programmer or free for writing those graphics extensions maybe.  Of course, if you have a detailed knowledge of how zero page addresses are used by EHBASIC, you might be able to get away with re-using some of the existing locations for new functions.  I never felt safe enough to risk it.

The reason I am talking about this at all is that there are other discussions about extending ehBasic to support save / load and graphics functions.  In all likelihood, those functions will need some zero page of their own.  So it's useful to know where the BASICZP and BASICZP0 (in my example above) actually begin and end so you know what other addresses in ZP you can safely use.

**Reply by netzerohero** [reply]
@linuxplayground 
Thank you for your insightful and helpful response. 

` have it laid out [zero-page] in two segments. One for stuff that must live at the bottom of zero page and another that can be relocatable within zero page. `

I understand and agree with your comments regarding EhBasic's use of zero-page; hence my attempt to capture memory map in a spreadsheet -- to help with memory usage visualization -- and augment source code's definition currently in ehbasic/src/zp.inc.

As currently implemented on rp6502 picocomputer, when EhBASIC runs, it fundamentally "owns" the machine.  (Like the BASIC machines of 50 years ago.)  It currently doesn't need to "play nice" with any OS, it hasn't had a need to co-exist with any c-library functions or c-runtime.  But should feature enhancement to EhBASIC go that direction, as your insights suggest -- well that is why I put the two RAM-environments side-by-side in the memory-map spreadsheet: C-environment and EhBASIC.  (It is understood zero-page and page-1 as well as FF-page have special usage on 6502 hw.)

I'll update the memory-map spreadsheet following your helpful comments. Thank you.
[edit: updated: 20231210_1119]

**Comment by netzerohero** [comment]
Updated memory-map spreadsheet available above - following Rumble's [15-Dec-2023 EhBASIC's memory footprint improvements](https://github.com/orgs/picocomputer/discussions/59#discussioncomment-7869541).
[updated: 20231216_1744]
